docs/proposals/error-handling-guideline.md

```diff
@@ -0,0 +1,279 @@
+- Provide specific and actionable prompt message with argument suggestion or show the example usage for reference. (e.g, Instead of showing flag or argument options is missing, please provide available argument options and guide users to "--help" to view more examples)
+- If the actionable prompt message is too long to show in the CLI output, consider guide users to ORAS user guide or troubleshooting guide with the permanent link.
+- If the error message is not enough for troubleshooting, guide users to use "--verbose" to print much more detailed logs
+
+### Don'Ts
+
+- Do not use a formula-like or a programming expression in the error message. (e.g, `json: cannot unmarshal string into Go value of type map[string]map[string]string.`, or `Parameter 'xyz' must conform to the following pattern: '^[-\\w\\._\\(\\)]+$'`)
+- Do not use ambiguous expressions which mean nothing to users. (e.g, `Something unexpected happens`, or `Error: accepts 2 arg(s), received 0`)
+- Do not print irrelevant error message to make the output noisy. The more irrelevant output you produce, the longer it's going to take the user to figure out what they did wrong.
+
+## How to write friendly error message
+
+### Recommended error message structure
+
+Here is a sample structure of an error message:
+
+```text
+Error: {Error|Error response from registry}: {Error description}
+[Usage: {Command usage}]
+[{Recommended solution}]
+```
+
+- Status code is an optional information. Printed out the status code if the error message is generated from the server side. 
+- Command usage is also an optional information but it's recommended to be printed out when user input doesn't follow the standard usage or examples.
+- Recommended solution (if any) should follow the general guiding principles described above.
+
+### Examples
+
+Here are some examples of writing error message with helpful prompt actionable information:
+
+#### Example 1: When no reference provided in `oras copy`
+
+Current behavior and output:
+
+```
+$ oras cp
+Error: accepts 2 arg(s), received 0
+```
+
+Suggested error message:
+
+```
+$ oras cp
+Error: "oras copy" requires exactly 2 arguments but received 0.
+Usage: oras copy [flags] <from>{:<tag>|@<digest>} <to>[:<tag>[,<tag>][...]]
+Please specify 2 arguments as source and destination respectively. Run "oras copy -h" for more options and examples
+```
+
+#### Example 2: When reference does not match with the expected format
+
+Current behavior and output:
+
+```
+$ oras tag list ghcr.io/oras-project/oras
+Error: unable to add tag for 'list': invalid reference: missing repository
+```
+
+Suggested error message:
+
+```
+$ oras tag list ghcr.io/oras-project/oras
+Error: There is no "list" sub-command for "oras tag" command.
+Usage: oras tag [flags] <name>{:<tag>|@<digest>} <new_tag> [...]
+If you want to tag a manifest in a registry or an OCI image layout, use "oras tag". If you want to list available tags in a repository, use "oras repo show tags" 
```

Comment:
The users clearly typed `oras tag list`. Why do we need to mention `oras tag`? It seems not related.

Reply:
Agree. Removed the first sentence from the prompt information.